File: smallestInfiniteSet.java
import java.util.*;
// Approach 1- Heap + Set
class Solution { // TC O ((m+n). log n) SC O(n)
    private HashSet<Integer> isPresent;
    private PriorityQueue<Integer> addedIntegers;
    private Integer currentInteger;
    
    public Solution() {
        isPresent = new HashSet<>();
        addedIntegers = new PriorityQueue<>();
        currentInteger = 1;
    }
    
    public int popSmallest() {
        int answer;
        // If there are numbers in the min-heap, 
        // top element is lowest among all the available numbers.
        if (!addedIntegers.isEmpty()) {
            answer = addedIntegers.poll();
            isPresent.remove(answer);
        } 
        // Otherwise, the smallest number of large positive set 
        // denoted by 'currentInteger' is the answer.
        else {
            answer = currentInteger;
            currentInteger += 1;
        }
        return answer;
    }
    
    public void addBack(int num) {
        if (currentInteger <= num || isPresent.contains(num)) {
            return;
        }
        // We push 'num' in the min-heap if it isn't already present.
        addedIntegers.add(num);
        isPresent.add(num);
    }
}

// Approach 2- TreeSet 
class Solution1{ // TC O ((m+n). log n) SC O(n)
    private Set<Integer> set;
    private int currentInteger;

    public Solution1(){
        set=new TreeSet<>();
        currentInteger=1;
    }

    public int popSmallest(){
        int answer;
        if(!set.isEmpty()){
            answer=set.first();
            set.remove(answer);
        } else {
            answer=currentInteger;
            currentInteger++;
        }
        return answer;
    }

    public void addBack(int num){
        if(currentInteger<=num || set.contains(num)) return;

        set.add(num);
    }

}

class smallestInfiniteSet{
    public static void main(String[] args) {
        Solution s=new Solution();
        s.addBack(2);
        System.out.println(s.popSmallest());
        Solution1 s1=new Solution1();
        s.addBack(2);
        System.out.println(s1.popSmallest());
    }
}

File: smallestInfiniteSet.py
class SmallestInfiniteSet:
# Approach 1- Heap + Set
    def __init__(self): # TC O ((m+n). log n) SC O(n)
        self.is_present: {int} = set()
        self.added_integers: [int] = []
        self.current_integer = 1

    def popSmallest(self) -> int:
        # If there are numbers in the min-heap, 
        # top element is lowest among all the available numbers.
        if len(self.added_integers):
            answer = heapq.heappop(self.added_integers)
            self.is_present.remove(answer)
        # Otherwise, the smallest number of large positive set 
        # denoted by 'current_integer' is the answer.
        else:
            answer = self.current_integer
            self.current_integer += 1
        return answer

    def addBack(self, num: int) -> None:
        if self.current_integer <= num or num in self.is_present:
            return
        # We push 'num' in the min-heap if it isn't already present.
        heapq.heappush(self.added_integers, num)
        self.is_present.add(num)

from sortedcontainers import SortedSet

class SmallestInfiniteSet: # TC O ((m+n). log n) SC O(n)
    def __init__(self):
        self.added_integers = SortedSet()
        self.current_integer = 1
    def popSmallest(self) -> int:
        # If there are numbers in the sorted-set, 
        # top element is lowest among all the available numbers.
        if len(self.added_integers):
            answer = self.added_integers[0]
            self.added_integers.discard(answer)
        # Otherwise, the smallest number of large positive set 
        # denoted by 'current_integer' is the answer.
        else:
            answer = self.current_integer
            self.current_integer += 1
        return answer
    def addBack(self, num: int) -> None:
        if self.current_integer <= num or num in self.added_integers:
            return
        # We push 'num' in the sorted-set if it isn't already present.
        self.added_integers.add(num)